### PR TITLE
Feat: enable Payment Stream cancellation via a Motion

### LINKED ISSUE #2216: Streaming Payments API: Cancelling streams using motions
Parent issue: https://github.com/JoinColony/colonyCDapp/issues/2209

- New saga needs to be implemented
- Relevant block-ingestor handlers need to be implemented
- A corresponding colony action should be created
- It should be possible to annotate the transaction

### PR DESCRIPTION
**Pull request by rumzledz**
## Description

Here are the required block-ingestor updates: [Feat: enable handler for the cancel streaming payment motion](https://github.com/JoinColony/block-ingestor/pull/240)

## Testing

1. Install the "Reputation Weighted" extension and use the "Testing governance" preset
2. Install the "Streaming Payments" extension
3. Using the debugging interface, create a Streaming Payment
Screenshot 2024-06-14 at 10 56 11
4. Wait for 3-5 seconds, then click the "Use latest Streaming Payment Native ID" button
Screenshot 2024-06-14 at 10 57 42
5. Copy the new value in the Streaming Payments input field, this should be the Native ID of the Streaming Payment you just made
6. Open the GraphQL playground http://localhost:20002/ and run the following query:
> Replace NATIVE_ID with the Native ID you just copied.
> This query should show you the Streaming Payment you just made.
```gql
query ListStreamingPayments {
  listStreamingPayments(filter: {nativeId: {eq: NATIVE_ID}}) {
    items {
      isCancelled
    }
  }
}
```
7. Verify that the `isCancelled` field is set to `null`
Screenshot 2024-06-17 at 07 37 55
8. Click the "Cancel via motion" button
Screenshot 2024-06-14 at 11 01 52
9. Wait for 3-5 seconds then click this button to open the action you just made, then copy the Transaction Hash from the URL
Screenshot 2024-06-17 at 07 31 47
10. Run the following query:
> Replace TRANSACTION_HASH with the Transaction Hash you just copied.
> This should show you both the Streaming Payment you made and the motion you made
```gql
query ListStreamingPayments {
  listStreamingPayments(filter: {nativeId: {eq: NATIVE_ID}}) {
    items {
      isCancelled
    }
  }
  listColonyMotions(
    filter: {transactionHash: {eq: TRANSACTION_HASH}}
  ) {
    items {
      isFinalized
    }
  }
}
```
11. Make sure that:
- The Streaming Payment's `isCancelled` field is set to `null`
- The motion's `isFinalized` field is set to `false`
Screenshot 2024-06-14 at 10 53 02
12. Open the User Hub transaction entry for the "Cancel Streaming Payment" motion you started
13. Approve the motion and back it up with the maximum stake
14. Advance the time by running the following commands:
```bash
curl -X POST --data '{"jsonrpc":"2.0","method":"evm_increaseTime","params":[3600],"id":1}' localhost:8545
curl -X POST --data '{"jsonrpc":"2.0","method":"evm_mine","params":[],"id":1}' localhost:8545
```
15. Reload the page
16. Finalise the motion
17. Run the query on step 10 to get an update for your Streaming Payment and the motion you made
18. Verify that:
- The Streaming Payment's `isCancelled` field is set to `true`
- The motion's `isFinalized` field is set to `true`
Screenshot 2024-06-14 at 10 53 55

## Diffs

**New stuff** ✨

* A new saga for handling the "Cancel Streaming Payment" motion

Resolves #2216

### CODE REVIEW

File: src/components/v5/payments/TmpAdvancedPayments.tsx [hunk 1]
```diff
@@ -98,6 +106,20 @@ const TmpAdvancedPayments = () => {
...
+    gql`
+      query ListStreamingPayments {
+        listStreamingPayments {
+          nextToken
+          items {
+            nativeId
+          }
+        }
+      }
+    `,
```
Comment (jakubcolony) [inline src/components/v5/payments/TmpAdvancedPayments.tsx:119]: This is okay since it's a testing UI, otherwise it should live in a `.graphql` file

File: src/redux/sagas/motions/expenditures/cancelStreamingPaymentsMotion.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    );
+
+    yield createGroupTransaction(createMotion, batchKey, meta, {
+      context: ClientType.VotingReputationClient,
+      methodName: 'createMotion',
+      identifier: colonyAddress,
+      params: [
+        motionDomainId,
+        childSkillIndex,
+        streamingPaymentsClient.address,
```
Comment (jakubcolony) [inline src/redux/sagas/motions/expenditures/cancelStreamingPaymentsMotion.ts:113]: 😉

File: src/components/v5/payments/TmpAdvancedPayments.tsx [hunk 3]
```diff
@@ -533,45 +577,89 @@ const TmpAdvancedPayments = () => {
...
+                    'bg-base-white/20 pointer-events-none absolute flex h-full w-full items-center justify-center opacity-0 backdrop-blur-sm transition-opacity',
+                    {
+                      'opacity-100': copiedMsgVisible,
+                    },
+                  )}
+                >
+                  <p className="font-bold">Copied to clipboard! 😎</p>
+                </div>
+              </button>
+            )}
```
Comment (jakubcolony) [inline src/components/v5/payments/TmpAdvancedPayments.tsx:660]: I liked the ease of having clickable links to a motion logged in the console, do you think this could redirect to the action with latest tx hash instead?

**Review by jakubcolony — APPROVED** [review]
Well done with pinpointing the finalization issues, it works flawlessly now:

https://github.com/JoinColony/colonyCDapp/assets/112586815/5070513e-ae99-44a6-8b6e-7e81fdf9eda5

Can also see the related action(s):
image

File: src/components/v5/payments/TmpAdvancedPayments.tsx (hunk 1 above)
Comment (rumzledz) [reply src/components/v5/payments/TmpAdvancedPayments.tsx:119]: My thoughts exactly, I didn't wanna add this to the `streamingPayments.graphql` file:
```gql
query ListStreamingPayments {
  listStreamingPayments {
    items {
      ...StreamingPayment
    }
  }
}
```
Because I didn't wanna pollute the `generated.ts` file with a new query, lazy query etc. that won't be used in prod, at least from the look of things!

File: src/components/v5/payments/TmpAdvancedPayments.tsx (hunk 3 above)
Comment (rumzledz) [reply src/components/v5/payments/TmpAdvancedPayments.tsx:660]: On it!

File: docker/colony-cdapp-dev-env-orchestration.yaml [hunk 4]
```diff
@@ -88,7 +88,7 @@ services:
     image: colony-cdapp-dev-env/base:latest
     volumes:
       - 'amplify-mock-data:/mock-data'
-    command: "npx --yes http-server --cors --port 3006 /mock-data/colonyNetworkArtifacts"
+    command: 'npx --yes http-server --cors --port 3006 /mock-data/colonyNetworkArtifacts'
```
Comment (iamsamgibbs) [inline docker/colony-cdapp-dev-env-orchestration.yaml:91]: Unnecessary change?

**Review by iamsamgibbs — APPROVED** [review]
Looks good to me! All working as expected.

Screenshot 2024-06-17 at 13 18 15

File: docker/colony-cdapp-dev-env-orchestration.yaml (hunk 4 above)
Comment (rumzledz) [reply docker/colony-cdapp-dev-env-orchestration.yaml:91]: It was bugging me that the rest of the strings in this yaml file are single-quoted 😕 Should I revert it @iamsamgibbs?

Comment (iamsamgibbs) [reply docker/colony-cdapp-dev-env-orchestration.yaml:91]: Oh that is fine with me! My brain saw these the wrong way round and thought you were making it not match. 🙃

**Review by rdig — APPROVED** [review]
One thing that stuck out, but otherwise this is functioning as expected.

Tested your flow which seems to be working as you described:

Screenshot from 2024-06-18 00-23-34
Screenshot from 2024-06-18 00-23-41
Screenshot from 2024-06-18 00-24-10
Screenshot from 2024-06-18 00-24-46
Screenshot from 2024-06-18 00-25-21
Screenshot from 2024-06-18 00-30-08
Screenshot from 2024-06-18 00-30-25

File: src/redux/sagas/motions/expenditures/cancelStreamingPaymentsMotion.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+      !streamingPayment
+    ) {
+      throw new Error('Invalid payload');
+    }
+
+    const batchKey = 'cancelStreamingPaymentMotion';
+
+    const { colonyAddress } = colony;
+
+    if (colony.version < 15) {
```
[the remark this answers is not in the document]
Comment (rumzledz) [reply src/redux/sagas/motions/expenditures/cancelStreamingPaymentsMotion.ts:55]: Oh right, I shall find this exported value and will update in other sagas as well that are using this check 👌
